src/conduit-lwt-unix/conduit_lwt_unix_ssl.real.ml

```diff
@@ -52,22 +52,59 @@ module Client = struct
 
   let default_ctx = create_ctx ()
 
-  let connect ?(ctx = default_ctx) ?src ?hostname sa =
+  type verify = { hostname : bool; ip : bool }
+
+  let default_verify = { hostname = true; ip = false }
+
+  let verification { hostname; ip } = function
+    | None, _ when hostname -> invalid_arg "impossible to verify hostname"
+    | _, None when ip -> invalid_arg "impossible to verify ip"
+    | h, i ->
+        let hostname = if hostname && h <> None then h else None in
+        let ip = if ip && i <> None then i else None in
+        (hostname, ip)
+
+  let validate_hostname hostname =
+    try
+      let _ = Domain_name.(host_exn (of_string_exn hostname)) in
+      hostname
+    with Invalid_argument msg ->
+      let s =
+        Printf.sprintf "couldn't convert %s to a [`host] Domain_name.t: %s"
+          hostname msg
+      in
+      invalid_arg s
+
+  let connect ?(ctx = default_ctx) ?src ?hostname ?ip ?verify sa =
     Conduit_lwt_server.with_socket sa (fun fd ->
         (match src with
         | None -> Lwt.return_unit
         | Some src_sa -> Lwt_unix.bind fd src_sa)
         >>= fun () ->
         Lwt_unix.connect fd sa >>= fun () ->
+        let verify = Option.value ~default:default_verify verify in
+        let with_socket f =
+          let s = Lwt_ssl.embed_uninitialized_socket fd ctx in
+          let socket = Lwt_ssl.ssl_socket_of_uninitialized_socket s in
+          f socket;
+          Lwt_ssl.ssl_perform_handshake s
+        in
+        let maybe_verify ssl = function
+          | Some hostname, Some ip ->
+              Ssl.set_hostflags ssl [ Ssl.No_partial_wildcards ];
+              Ssl.set_host ssl (validate_hostname hostname);
+              Ssl.set_ip ssl (Ipaddr.to_string ip)
+          | Some hostname, None ->
+              Ssl.set_hostflags ssl [ Ssl.No_partial_wildcards ];
+              Ssl.set_host ssl (validate_hostname hostname)
+          | None, Some ip -> Ssl.set_ip ssl (Ipaddr.to_string ip)
+          | None, None -> ()
+        in
         (match hostname with
         | Some host ->
-            let s = Lwt_ssl.embed_uninitialized_socket fd ctx in
-            let ssl = Lwt_ssl.ssl_socket_of_uninitialized_socket s in
-            Ssl.set_client_SNI_hostname ssl host;
-            (* Enable hostname verification *)
-            Ssl.set_hostflags ssl [ Ssl.No_partial_wildcards ];
-            Ssl.set_host ssl host;
-            Lwt_ssl.ssl_perform_handshake s
+            with_socket (fun ssl ->
+                Ssl.set_client_SNI_hostname ssl host;
+                maybe_verify ssl (verification verify (hostname, ip)))
         | None -> Lwt_ssl.ssl_connect fd ctx)
```

Comment:
shouldn't here in the none case, when no hostname is provided, still the ip verificaton been done?

Reply:
"hostname" is a misnomer, it is a string that represents a remote host, it can easily be an IP.

I made the option to keep the previous behaviour of sending it as an SNI.

Reply:
My remark was not about SNI or not (which, if `hostname` is an IP, shouldn't be set), but that in the `hostname = None` case there's no call to `maybe_verify` which I'd have expected, but only a call to `Lwt_ssl.ssl_connect` -- so what kind of verification (if any) is performed here?

Of course, if you're not interested in that code path, leave it as is. It would be great if you could formulate an entry for CHANGES: which behaviour changed and why.

Reply:
> if hostname is an IP, shouldn't be set)

Will change it, I was being conservative and retaining previous behaviour

> there's no call to maybe_verify which I'd have expected

I see what you mean, I'll change it to check the IP if the setting is enabled, in general I don't think it should be done as few certificates contain an IP as the subject.

> It would be great if you could formulate an entry for CHANGES: which behaviour changed and why.

Sure thing

Reply:
> > if hostname is an IP, shouldn't be set)
>
> Will change it, I was being conservative and retaining previous behaviour

I don't know for sure how OpenSSL behaves, but in the TLS RFC it is very clear that setting the server name identification to an IP address is invalid. It must be a valid hostname. If the current patch works fine for you in such a scenario, feel free to leave it as is.

> I see what you mean, I'll change it to check the IP if the setting is enabled, in general I don't think it should be done as few certificates contain an IP as the subject.

Well, most DNS-over-TLS certificates include an IP address. But again, if you think it is not worth it, I'm fine to merge as is.

Reply:
Now it shouldn't send the SNI when a DNS hostname isn't available and it should be able to verify the IP in the certificate if the users wants it to, let me know if the changelog looks good